lib/iris/_concatenate.py

```diff
@@ -917,97 +1062,79 @@ def register(
             elif not match:
                 mismatch_error_msg = f"Found cubes with overlap on concatenate axis {candidate_axis}, skipping concatenation for these cubes"
 
-        # Check for compatible AuxCoords.
-        if match:
-            if check_aux_coords:
-                for coord_a, coord_b in zip(
-                    self._cube_signature.aux_coords_and_dims,
-                    cube_signature.aux_coords_and_dims,
+        def get_hash(array):
+            return hashes[array_id(array)]
+
+        def get_hashes(coord):
+            result = []
+            if hasattr(coord, "core_points"):
+                result.append(get_hash(coord.core_points()))
+                if coord.has_bounds():
+                    result.append(get_hash(coord.core_bounds()))
+            else:
+                result.append(get_hash(coord.core_data()))
```

Comment:
A added a case where this raises a TypeError in 470f28e, does that work for you?

Reply:
With the added typehints, the TypeError seemed unnecessary because the code is unreachable if users respect the type hints. Since type checking is enabled and the function is not part of an external API, users will likely respect the type hints. Therefore I removed the TypeError again in f481d55. However, if you prefer, I can put it back?

Reply:
Good thought - with MyPy in place we should be able to gradually remove defensive code. My original suggestion was only based on my experience with linting, anyway (PyCharm), rather than any concern that the code might actually crash - it's all internal and tightly controlled.